underfs/gsg/src/main/java/alluxio/underfs/obs/GSGUnderFileSystem.java

```diff
@@ -0,0 +1,299 @@
+    mStorageClient = storageClient;
+    mBucketClient = bucketClient;
+    mBucketName = bucketName;
+  }
+
+  @Override
+  public String getUnderFSType() {
+    return "GSG";
+  }
+
+  // Setting GSG owner via Alluxio is not supported yet. This is a no-op.
+  @Override
+  public void setOwner(String path, String user, String group) {}
+
+  // Setting GSG mode via Alluxio is not supported yet. This is a no-op.
+  @Override
+  public void setMode(String path, short mode) {}
+
+  @Override
+  protected boolean copyObject(String src, String dst) {
+    LOG.debug("Copying {} to {}", src, dst);
+    Storage.CopyRequest request = Storage.CopyRequest.newBuilder()
+        .setSource(BlobId.of(mBucketName, src))
+        .setTarget(BlobId.of(mBucketName, dst))
+        .build();
+    try {
+      Blob blob = mStorageClient.copy(request).getResult();
+      if (blob != null) {
+        return true;
+      }
+    } catch (StorageException e) {
+      LOG.error("Failed to copy file {} to {}", src, dst, e);
+    }
+    return false;
+  }
+
+  @Override
+  public boolean createEmptyObject(String key) {
+    try {
+      BlobId blobId = BlobId.of(mBucketName, key);
+      BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+      Blob blob = mStorageClient.create(blobInfo);
+      if (blob == null) {
+        LOG.error("Failed to create object {}", key);
+        return false;
+      }
+    } catch (StorageException e) {
+      LOG.error("Failed to create object: {}", key, e);
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  protected OutputStream createObject(String key) throws IOException {
+    return new GSGOutputStream(mBucketName, key, mStorageClient);
+  }
+
+  @Override
+  protected boolean deleteObject(String key) {
+    // TODO(lu) GSG supports bunch delete
+    BlobId blobId = BlobId.of(mBucketName, key);
+    try {
+      if (!mStorageClient.delete(blobId)) {
+        LOG.error("Failed to delete object {}", key);
+        return false;
+      }
+    } catch (StorageException e) {
+      LOG.error("Failed to delete {}", key, e);
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  protected String getFolderSuffix() {
+    return FOLDER_SUFFIX;
+  }
+
+  @Override
+  protected ObjectListingChunk getObjectListingChunk(String key, boolean recursive) throws IOException {
+    key = PathUtils.normalizePath(key, PATH_SEPARATOR);
+    // In case key is root (empty string) do not normalize prefix
+    key = key.equals(PATH_SEPARATOR) ? "" : key;
+    Page<Blob> blobPages = getObjectListingChunk(key);
+    if (blobPages != null && blobPages.getValues().iterator().hasNext()) {
+      return new GSGObjectListingChunk(blobPages, Pattern.compile("^" + key + "([^/]+)/?$"), recursive);
+    }
+    return null;
+  }
+
+  // Get next chunk of listing result.
+  private Page<Blob> getObjectListingChunk(String key) throws IOException {
+    try {
+      return mBucketClient.list(Storage.BlobListOption.prefix(key));
```

Comment:
Based on recursive we may want to specify the `currentDirectory` option
For optimization (later) we can also used the `Fields` option to limit the returned data size.

Reply:
Thanks a lot!!
I changed to use `mBucketClient.list(Storage.BlobListOption.prefix(key), Storage.BlobListOption.currentDirectory());` when `recursive=false`.